File: plugins/transformometro/src/utils/brazilNationalHolidays.ts
const holidayCache = new Map<number, Set<string>>();

function toIso(d: Date): string {
  const year = d.getFullYear();
  const month = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function addDays(base: Date, days: number): Date {
  const copy = new Date(base.getFullYear(), base.getMonth(), base.getDate());
  copy.setDate(copy.getDate() + days);
  return copy;
}

/** Domingo de Páscoa (calendário gregoriano). */
export function easterSunday(year: number): Date {
  const a = year % 19;
  const b = Math.floor(year / 100);
  const c = year % 100;
  const d = Math.floor(b / 4);
  const e = b % 4;
  const f = Math.floor((b + 8) / 25);
  const g = Math.floor((b - f + 1) / 3);
  const h = (19 * a + b - d - g + 15) % 30;
  const i = Math.floor(c / 4);
  const k = c % 4;
  const l = (32 + 2 * e + 2 * i - h - k) % 7;
  const m = Math.floor((a + 11 * h + 22 * l) / 451);
  const month = Math.floor((h + l - 7 * m + 114) / 31);
  const day = ((h + l - 7 * m + 114) % 31) + 1;
  return new Date(year, month - 1, day);
}

export function brazilNationalHolidaySet(year: number): Set<string> {
  const cached = holidayCache.get(year);
  if (cached) return cached;

  const easter = easterSunday(year);
  const holidays = new Set<string>([
    `${year}-01-01`,
    `${year}-04-21`,
    `${year}-05-01`,
    `${year}-09-07`,
    `${year}-10-12`,
    `${year}-11-02`,
    `${year}-11-15`,
    `${year}-12-25`,
    toIso(addDays(easter, -48)),
    toIso(addDays(easter, -47)),
    toIso(addDays(easter, -2)),
    toIso(addDays(easter, 60)),
  ]);

  holidayCache.set(year, holidays);
  return holidays;
}

export function isNationalHoliday(date: Date): boolean {
  const iso = toIso(date);
  return brazilNationalHolidaySet(date.getFullYear()).has(iso);
}

File: transformometro-api/tm_app/core/brazil_national_holidays.py
"""Feriados nacionais fixos e móveis (Brasil) para cálculo de dias úteis."""

from __future__ import annotations

from datetime import date, timedelta
from functools import lru_cache


def easter_sunday(year: int) -> date:
    """Domingo de Páscoa (calendário gregoriano)."""
    a = year % 19
    b = year // 100
    c = year % 100
    d = b // 4
    e = b % 4
    f = (b + 8) // 25
    g = (b - f + 1) // 3
    h = (19 * a + b - d - g + 15) % 30
    i = c // 4
    k = c % 4
    l = (32 + 2 * e + 2 * i - h - k) % 7
    m = (a + 11 * h + 22 * l) // 451
    month = (h + l - 7 * m + 114) // 31
    day = ((h + l - 7 * m + 114) % 31) + 1
    return date(year, month, day)


@lru_cache(maxsize=64)
def brazil_national_holidays(year: int) -> frozenset[date]:
    """Feriados nacionais segundo calendário federal (fixos + móveis ligados à Páscoa)."""
    easter = easter_sunday(year)
    return frozenset(
        {
            date(year, 1, 1),
            date(year, 4, 21),
            date(year, 5, 1),
            date(year, 9, 7),
            date(year, 10, 12),
            date(year, 11, 2),
            date(year, 11, 15),
            date(year, 12, 25),
            easter - timedelta(days=48),
            easter - timedelta(days=47),
            easter - timedelta(days=2),
            easter + timedelta(days=60),
        }
    )


def is_national_holiday(value: date) -> bool:
    return value in brazil_national_holidays(value.year)
